### PR TITLE
Enable flag to not pass PAD tokens in ffwd

### PR DESCRIPTION
**Pull request by bcui19**
This PR does two things:

1. Modifies the `attn_bias` function to always return the attention_mask.
2. Enables us to remove pad tokens before calling `.forward` on the ffwd network then re-add in the pad tokens.

Loss curves on a fully randomly initialized network:
image

We also get slightly higher throughput from this when there are PAD tokens in our dataset (and no degradation when compared to `main` with `attn_impl: triton`:
image

wandb: https://wandb.ai/mosaic-ml/padding_check?workspace=user-bcui

### CODE REVIEW

File: llmfoundry/models/layers/ffn_padding_utils.py
Comment (vchiley) [inline llmfoundry/models/layers/ffn_padding_utils.py:1]: can this be imported from `https://github.com/sz128/flash-attention/blob/3c0d49532cb3bc0a36b405f34590ff82cf582853/flash_attn/bert_padding.py`?

File: llmfoundry/models/layers/blocks.py
Comment (vchiley) [inline llmfoundry/models/layers/blocks.py:1]: Does passing `attention_mask` into attn fwd break anything?
I don't think it should, just asking.

Comment (vchiley) [reply llmfoundry/models/layers/blocks.py:1]: https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/models/layers/attention.py#L436

this gets triggered

Comment (vchiley) [reply llmfoundry/models/layers/blocks.py:1]: https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/models/layers/attention.py#L157
This is triggered in torch attn impl

**Comment by vchiley** [comment]
> We also get slightly higher throughput from this when there are PAD tokens in our dataset:

Can you run main vs your branch with use_pad_tok_in_ffwd flag vs your branch without use_pad_tok_in_ffwd flag?

File: llmfoundry/models/layers/__init__.py [hunk 1]
```diff
@@ -9,24 +9,14 @@
...
-    'FC_CLASS_REGISTRY',
-    'SharedEmbedding',
-    'FFN_CLASS_REGISTRY',
-    'build_ffn',
+    'scaled_multihead_dot_product_attention', 'flash_attn_fn',
+    'triton_flash_attn_fn', 'MultiheadAttention', 'MultiQueryAttention',
+    'attn_bias_shape', 'build_attn_bias', 'build_alibi_bias',
+    'ATTN_CLASS_REGISTRY', 'MPTMLP', 'MPTBlock', 'NORM_CLASS_REGISTRY',
+    'LPLayerNorm', 'FC_CLASS_REGISTRY', 'SharedEmbedding', 'FFN_CLASS_REGISTRY',
+    'build_ffn', 'unpad_input', 'pad_input'
```
Comment (dakinggg) [inline llmfoundry/models/layers/__init__.py:21]: add trailing comma back so formatting stays nice

**Review by dakinggg — COMMENTED** [review]
Can you add a test that tests numerical equivalence of computation with and without the flag? might be off by a bit because of numerics, but lets see.

File: llmfoundry/models/layers/blocks.py
Comment (dakinggg) [reply llmfoundry/models/layers/blocks.py:1]: yeah, this is not ideal right? Any way we can avoid?

File: llmfoundry/models/layers/ffn_padding_utils.py
Comment (dakinggg) [reply llmfoundry/models/layers/ffn_padding_utils.py:1]: +1

File: llmfoundry/models/layers/blocks.py
Comment (bcui19) [reply llmfoundry/models/layers/blocks.py:1]: It's possible, I could just be checking the `attn_bias` for this value:`torch.finfo(attn_bias.dtype).min`

taken from here: https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/models/mpt/modeling_mpt.py#L418

@dakinggg would you prefer this?

Comment (dakinggg) [reply llmfoundry/models/layers/blocks.py:1]: I think that works? @vchiley ?

File: llmfoundry/models/layers/blocks.py [hunk 2]
```diff
@@ -132,6 +135,22 @@ def forward(
...
+                    'Please install flash-attn==1.0.9 or flash-attn==2.3.2')
+            m, indices, _, _ = unpad_input(m, attention_mask)
         n = self.ffn(m)
+        if not self.use_pad_tok_in_ffwd:
+            try:
+                from flash_attn.bert_padding import pad_input  # type: ignore # yapf: disable # isort: skip
+            except:
+                raise RuntimeError(
+                    'Please install flash-attn==1.0.9 or flash-attn==2.3.2')
+            n = pad_input(n, indices, batch_size, seq_len)
```
Comment (vchiley) [inline llmfoundry/models/layers/blocks.py:154]: I'd put
```
try:
    from flash_attn.bert_padding import unpad_input, pad_input  # type: ignore # yapf: disable # isort: skip
except:
    unpad_input, pad_input = None, None
```
at the top of the file

then have this as
```suggestion
        batch_size = m.size(0)
        seq_len = m.size(1)
        if not self.use_pad_tok_in_ffwd:
            if unpad_input is None:
                raise RuntimeError(
                    'Please install flash-attn==1.0.9 or flash-attn==2.3.2')
            m, indices, _, _ = unpad_input(m, attention_mask)
        n = self.ffn(m)
        if not self.use_pad_tok_in_ffwd:
            n = pad_input(n, indices, batch_size, seq_len)
```
(removed the second runtime error cuz if the first one passed, the second one should be fine)

File: llmfoundry/models/layers/blocks.py
Comment (dakinggg) [reply llmfoundry/models/layers/blocks.py:1]: Ok, I leave it up to you which method you prefer.

Comment (vchiley) [reply llmfoundry/models/layers/blocks.py:1]: its either we use BW to find `attn_bias.min()` or we use BW to add a mask to `attn_bias`
these use approximately the same amount of BW... 🤷‍♂️ 

I'd say leave as is

Comment (bcui19) [reply llmfoundry/models/layers/blocks.py:1]: With `attn_impl: triton` we have the same loss and same throughput:

image
image

File: llmfoundry/models/layers/blocks.py [hunk 3]
```diff
@@ -132,6 +140,15 @@ def forward(
         m = x
         if self.norm_2 is not None:
             m = self.norm_2(x)
+        batch_size = m.size(0)
+        seq_len = m.size(1)
```
Comment (mvpatel2000) [inline llmfoundry/models/layers/blocks.py:144]:
```suggestion
        batch_size, seq_len= m.size[:2]
```

**Review by mvpatel2000 — COMMENTED** [review]
Minor nits

File: llmfoundry/models/layers/blocks.py
Comment (dakinggg) [reply llmfoundry/models/layers/blocks.py:1]: I think this is ok. Its an optional flag that may or may not help performance, similar to many other configuration options.

Comment (dakinggg) [reply llmfoundry/models/layers/blocks.py:1]: Could you move the summary of performance differences into the PR description please? Also what setting did you originally see a throughput win on?

File: llmfoundry/models/layers/blocks.py [hunk 4]
```diff
@@ -132,6 +140,15 @@ def forward(
...
+        if not self.use_pad_tok_in_ffwd:
+            if unpad_input is None:
+                raise RuntimeError(
+                    'Please install flash-attn==1.0.9 or flash-attn==2.3.2')
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/models/layers/blocks.py:147]: I think this should actually be in `configuration_mpt.py`. There is a big ol `validate_config` function there, which is where this kind of check belongs.

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 5]
```diff
@@ -131,6 +132,7 @@ def __init__(
                 See llmfoundry.models.utils.param_init_fns.py for info on other param init config options
             fc_type (str): choose fc layer implementation. Options: torch and te. te layers support fp8 when using H100 GPUs.
             tie_word_embeddings (bool): Whether to tie the input embedding and output layers.
+            use_pad_tok_in_ffwd (bool): Whether to forward the pad token in the feedforwards networks.
```
Comment (dakinggg) [inline llmfoundry/models/mpt/configuration_mpt.py:135]:
```suggestion
            use_pad_tok_in_ffwd (bool): Whether to forward the pad token in the feedforward networks.
```

File: llmfoundry/models/layers/blocks.py
Comment (dakinggg) [reply llmfoundry/models/layers/blocks.py:1]: oh wait, sorry, the attention mask thing makes it slower when you're _not_ using the new option? @vchiley wdyt about this?

Comment (bcui19) [reply llmfoundry/models/layers/blocks.py:1]: Yeah but only with `attn_impl: torch,` with `attn_impl: triton` it's faster

Comment (vchiley) [reply llmfoundry/models/layers/blocks.py:1]: i mean no one should be using the torch path
thats more for testing 

so long as the triton path isn't effected, we're good.

File: llmfoundry/models/layers/blocks.py [hunk 6]
```diff
@@ -53,6 +58,7 @@ def __init__(
         fc_type: str = 'torch',
         device: Optional[str] = None,
         no_bias: bool = False,
+        use_pad_tok_in_ffn: bool = True,
```
Comment (vchiley) [inline llmfoundry/models/layers/blocks.py:61]:
```suggestion
        use_pad_tok_in_ffn: bool = False,
```
default False

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 7]
```diff
@@ -60,6 +60,7 @@ def __init__(
         init_config: Dict = init_config_defaults,
         fc_type: str = 'torch',
         tie_word_embeddings: bool = True,
+        use_pad_tok_in_ffn: bool = True,
```
Comment (vchiley) [inline llmfoundry/models/mpt/configuration_mpt.py:63]:
```suggestion
        use_pad_tok_in_ffn: bool = False,
```

default False

File: tests/models/test_model.py [hunk 8]
```diff
@@ -831,6 +832,43 @@ def test_forward_with_padding(attention_impl: str, pos_emb_config: dict,
                 batched_output[1, :],
                 atol=1e-6 if attention_impl == 'torch' else 1e-8)
 
+        try:
```
Comment (vchiley) [inline tests/models/test_model.py:835]: having try except block in test seems counter-intuitive 🤔 

you probably cant import it on CPU tests. maybe gate by "if not running on gpu"?

File: tests/models/test_model.py [hunk 9]
```diff
@@ -731,6 +731,7 @@ def test_forward_with_padding(attention_impl: str, pos_emb_config: dict,
             'init_std': 0.02,
         },
         tie_word_embeddings=tie_word_embeddings,
+        use_pad_tok_in_ffn=True,
```
Comment (vchiley) [inline tests/models/test_model.py:734]: Isn't this the default?
Why do you need to explicitly set it?

**Review by dakinggg — APPROVED** [review]
Thanks!
